orchagent/aclorch.cpp

```diff
@@ -2447,10 +2464,123 @@ void AclOrch::doTask(Consumer &consumer)
+                                AclTable    &curT,
+                                set<string> &addSet,
+                                set<string> &delSet)
 {
+    set<string> newPortSet, curPortSet;
+
+    // Collect new ports
+    for (auto p : newT.pendingPortSet)
+    {
+        newPortSet.insert(p);
+    }
+    for (auto p : newT.portSet)
+    {
+        newPortSet.insert(p);
+    }
+
+    // Collect current ports
+    for (auto p : curT.pendingPortSet)
+    {
+        curPortSet.insert(p);
+    }
+    for (auto p : curT.portSet)
+    {
+        curPortSet.insert(p);
+    }
+
+    // Get the difference newPortSet-curPortSet and curPortSet-newPortsSet
+    std::set_difference(newPortSet.begin(), newPortSet.end(),
+                        curPortSet.begin(), curPortSet.end(),
+                        std::inserter(addSet, addSet.end()));
+    std::set_difference(curPortSet.begin(), curPortSet.end(),
+                        newPortSet.begin(), newPortSet.end(),
+                        std::inserter(delSet, delSet.end()));
+
+}
+
+bool AclOrch::updateAclTablePorts(AclTable &newTable, AclTable &curTable)
+{
+    sai_object_id_t    port_oid = SAI_NULL_OBJECT_ID;
+    set<string>        addPortSet, deletePortSet;
+
     SWSS_LOG_ENTER();
+    getAddDeletePorts(newTable, curTable, addPortSet, deletePortSet);
 
+    // Lets first unbind and unlink ports to be removed
+    for (auto p : deletePortSet)
+    {
+        SWSS_LOG_NOTICE("Deleting port %s from ACL list %s",
+                        p.c_str(), curTable.id.c_str());
+        if (curTable.pendingPortSet.find(p) != curTable.pendingPortSet.end())
+        {
+            SWSS_LOG_NOTICE("Removed:%s from pendingPortSet", p.c_str());
+            curTable.pendingPortSet.erase(p);
+        } else if (curTable.portSet.find(p) != curTable.portSet.end())
+        {
+            gPortsOrch->getAclBindPortId(p, port_oid);
+            assert(port_oid != SAI_NULL_OBJECT_ID);
+            assert(curTable.ports.find(port_oid) != curTable.ports.end());
+            if (curTable.ports[port_oid] != SAI_NULL_OBJECT_ID)
+            {
+                // Unbind and unlink
+                SWSS_LOG_NOTICE("Unbind and Unlink:%s", p.c_str());
+                curTable.unbind(port_oid);
+                curTable.unlink(port_oid);
+            }
+            SWSS_LOG_NOTICE("Removed:%s from portSet", p.c_str());
+            curTable.portSet.erase(p);
+        }
+    }
+
+    // Now link and bind ports to be added
+    for (auto p : addPortSet)
+    {
+        SWSS_LOG_NOTICE("Adding port %s to ACL list %s",
+                        p.c_str(), curTable.id.c_str());
+        Port port;
+        if (!gPortsOrch->getPort(p, port))
+        {
+            curTable.pendingPortSet.emplace(p);
+            continue;
+        }
+
+        if (!gPortsOrch->getAclBindPortId(p, port_oid))
+        {
+            throw runtime_error("updateAclTablePorts: Couldn't find portOID");
```

Comment:
No, here we expect the program to terminate and log the message we pass to rumtime_error() function. And thats the practice in orchagent that I have observed as well. Am I missing anything?

Reply:
@yxieca what do you think? I brought this up because in the past Ying has warned me about doing this in orchagent. I personally don't have strong feelings either way since it seems like this situation shouldn't ever really come up in practice, but I'll defer to him here.

Reply:
@vasant17 do you feel that throw exception and quit is the only option here? Is this never expect to happen in practice?

You observation of orchagent behavior is somewhat right but not totally up-to-date. We are making efforts to reduce exception quits and trying to handle errors gracefully.

If you could find a solution to the situation, please add handling code in, e.g. retry or wait util you have a valid port oid. If you strongly feel that this should never happen, please comment so.

Reply:
I don't see any code path where port object would be present and NOT OID. So, I will keep this for now with a comment.